src/transformers/models/efficientloftr/modeling_efficientloftr.py

```diff
@@ -0,0 +1,1345 @@
+            value_states,
+            current_attention_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            scaling=self.scaling,
+            **kwargs,
+        )
+
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        return attn_output, attn_weights
+
+
+class EfficientLoFTRMLP(nn.Module):
+    def __init__(self, config: EfficientLoFTRConfig):
+        super().__init__()
+        hidden_size = config.hidden_size
+        self.fc1 = nn.Linear(2 * hidden_size, 2 * hidden_size, bias=False)
+        self.activation = ACT2FN[config.mlp_activation_function]
+        self.fc2 = nn.Linear(2 * hidden_size, hidden_size, bias=False)
+        self.layer_norm = nn.LayerNorm(hidden_size)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.fc1(hidden_states)
+        hidden_states = self.activation(hidden_states)
+        hidden_states = self.fc2(hidden_states)
+        hidden_states = self.layer_norm(hidden_states)
+        return hidden_states
+
+
+def get_positional_embeddings_slice(
+    hidden_states: torch.Tensor, positional_embeddings: tuple[torch.Tensor, torch.Tensor]
+) -> tuple[torch.Tensor, torch.Tensor]:
+    batch_size, h, w, _ = hidden_states.shape
+    positional_embeddings = tuple(
+        tensor[:, :h, :w, :].expand(batch_size, -1, -1, -1) for tensor in positional_embeddings
+    )
+    return positional_embeddings
+
+
+class EfficientLoFTRAggregatedAttention(nn.Module):
+    def __init__(self, config: EfficientLoFTRConfig, layer_idx: int):
+        super().__init__()
+
+        self.q_aggregation_kernel_size = config.q_aggregation_kernel_size
+        self.aggregation = EfficientLoFTRAggregationLayer(config)
+        self.attention = EfficientLoFTRAttention(config, layer_idx)
+        self.mlp = EfficientLoFTRMLP(config)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        position_embeddings: Optional[tuple[torch.Tensor, torch.Tensor]] = None,
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> torch.Tensor:
+        batch_size, embed_dim, _, _ = hidden_states.shape
+
+        # Aggregate features
+        aggregated_hidden_states, attention_mask, encoder_hidden_states, encoder_attention_mask = self.aggregation(
+            hidden_states, attention_mask, encoder_hidden_states, encoder_attention_mask
+        )
+
+        attention_hidden_states = aggregated_hidden_states.reshape(batch_size, -1, embed_dim)
+        encoder_hidden_states = encoder_hidden_states.reshape(batch_size, -1, embed_dim)
+
+        if position_embeddings is not None:
+            position_embeddings = get_positional_embeddings_slice(aggregated_hidden_states, position_embeddings)
```

Comment:
this looks a bit messy, the layer producing them should already produce what you need